### PR TITLE
Save DAG parsing time if dag is not scheduleable

### PR DESCRIPTION
**Pull request by AutomationDev85**
Hi airflow community,
this is my fifth PR and be happy to work DAG parsing runtime. We faced an issue with slow dag parsing if DAG has huge amount of dagruns and DB query is slow down the run time during DAG parsing. Then the DAG parsing was skipped due to long running query and this is not required if DAG is not scheduled. I think DAGs  which are not scheduled have biggest change to have huge number dagruns and this save runtime during this use case.

This PR will check is the DAG is not scheduled and then skips the DB query which are not required to be executed.

### CODE REVIEW

File: airflow/models/dag.py
Comment (potiuk) [inline airflow/models/dag.py]: I think `any([not dag.isinstance(dag.timetable, NullTimeTable) for dag in dags])` syntax is slightly more readable, can you please change it this way?

Comment (AutomationDev85) [reply airflow/models/dag.py]: Yep you are right! I changed the code.

File: airflow/models/dag.py
Comment (potiuk) [inline airflow/models/dag.py]: Actually I think it is not what I asked for, but likely this is even more readable:

```suggestion
        if not all([dag.isinstance(dag.timetable, NullTimetable) for dag in dags]):
```

BTW. Can you also please add a unit test actually testing it?

Comment (jscheffl) [reply airflow/models/dag.py]: As PR #30706 changes the structure of ǸullTimeTable` and review feedback was pointing to not using `isinstance()` I propose to wait until #30706 is merged and then re-base this change accordingly.

Comment (AutomationDev85) [reply airflow/models/dag.py]: As PR https://github.com/apache/airflow/pull/30706 is merged I changed the logic accordingly and added a unit test for it.

File: airflow/models/dag.py
Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
        if any(dag.timetable.can_be_scheduled for dag in dags):
```

File: airflow/models/dag.py
Comment (uranusjr) [inline airflow/models/dag.py]: From past experience, Mypy will probably ask you to provide better type hints for these. Instead of having an empty dict here, it’s generally easier to add an else block.

Comment (jscheffl) [reply airflow/models/dag.py]: Adjusted code. Thanks for the hint!

File: airflow/models/dag.py
Comment (uranusjr) [inline airflow/models/dag.py]:
```suggestion
        # Skip these queries entirely if no DAGs can be scheduled to save time.
```

**Review by uranusjr — APPROVED** [review]
Just one suggestion to clarify a comment. Logic looks good to me.

**Review by potiuk — APPROVED** [review]

**Pull request merged**
